pkgs/development/tools/rust/cargo-bisect-rustc/0001-dynamically-patchelf-binaries.patch

```diff
@@ -0,0 +1,56 @@
+diff --git a/src/toolchains.rs b/src/toolchains.rs
+index 4d85e7c..b1353c6 100644
+--- a/src/toolchains.rs
++++ b/src/toolchains.rs
+@@ -259,6 +259,8 @@ impl Toolchain {
+             .map_err(InstallError::Download)?;
+         }
+
++        nix_patchelf(tmpdir.path().to_path_buf())
++            .expect("failed to patch toolchain for NixOS");
+         fs::rename(tmpdir.into_path(), dest).map_err(InstallError::Move)?;
+
+         Ok(())
+@@ -557,3 +559,42 @@ pub(crate) fn download_tarball(
+     }
+     download_tar_gz(client, name, &format!("{}.gz", url,), strip_prefix, dest)
+ }
++
++fn nix_patchelf(mut toolchain_path: PathBuf) -> Result<(), Error> {
++    toolchain_path.push("bin");
++
++    for entry in toolchain_path.read_dir()? {
++        let entry = entry?;
++        if !entry.file_type()?.is_file() {
++            continue;
++        }
++
++        eprintln!("info: you seem to be running NixOS. Attempting to patch {}",
++                  entry.path().to_str().unwrap());
++        let _ = ::std::process::Command::new("@patchelf@/bin/patchelf")
++            .arg("--set-interpreter")
++            .arg("@dynamicLinker@")
++            .arg(entry.path())
++            .output();
++    }
++
++    toolchain_path.pop();
++    toolchain_path.push("lib");
++
++    for entry in toolchain_path.read_dir()? {
++        let entry = entry?;
++        if !entry.file_type()?.is_file() {
++            continue;
++        }
++
++        eprintln!("info: you seem to be running NixOS. Attempting to patch {}",
++                  entry.path().to_str().unwrap());
++        let _ = ::std::process::Command::new("@patchelf@/bin/patchelf")
++            .arg("--set-rpath")
++            .arg("@libPath@")
++            .arg(entry.path())
++            .output();
```

Comment:
one implication of this is that you could patch something, update your system, do a `nix-collect-garbage` and now you can't use the patched binaries anymore as nix doesn't know that you want to retain them.

Reply:
That's correct, the same is true of nixpkgs' `rustup` package (and the `rustup-toolchain-install-master` package):

https://github.com/NixOS/nixpkgs/blob/5b1d8990bd02d643a08fb4f7f41e559ec42896c1/pkgs/development/tools/rust/rustup/default.nix#L33-L39

I don't think this is a problem though, `cargo-bisect-rustc` downloads lots of rustc toolchains to bisect bugs introduced in rustc, those toolchains are most often removed as the bisection progresses (or at the end, I forget exactly) - thus it is unlikely that we end up with broken toolchains from this package. Without this patch, `cargo-bisect-rustc` won't work on NixOS (without using `buildFHSUserEnv` at least), so having this package makes `cargo-bisect-rustc` practical to use on NixOS.

Reply:
I think in the longer term it would be nice to automatically GC root these, but I don't think this is too much of a problem for this particular usecase (similarly for rustup, I guess).